### PR TITLE
feat(rlm): add Recursive Language Model mode for large codebase review

### PR DESCRIPTION
**Pull request by anderskev**
## Summary

This PR implements improvements to the RLM (Recursive Language Model) code review system. Changes are informed by analysis of the RLM research paper (Khattab et al.) which documented similar challenges we encountered during implementation.

## What is an RLM?

Traditional code review loads the entire codebase into context, which becomes impractical for large repositories. Even frontier models exhibit **context rot** where quality degrades as context length increases. The paper demonstrates that the effective context window cannot be understood independently of task complexity—more complex problems degrade at shorter lengths.

**The key insight**: Long prompts should not be fed into the neural network directly but should instead be **treated as part of the environment** that the LLM can symbolically interact with.

```
┌─────────────────────────────────────────────────────────────────┐
│                    Traditional LLM Approach                      │
│  [Entire Codebase] ──────────────────────────────► [LLM Context] │
│       100K+ tokens                                  Context Rot! │
└─────────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────────────────────────────────┐
│                      RLM Approach                                │
│  [Codebase as Variable] ◄────► [Python REPL] ◄────► [LLM]       │
│       In Memory              Code Execution      Fresh Context   │
│                                                                  │
│  LLM writes code to:                                             │
│  1. Peek at structure (repo.files, repo.structure)               │
│  2. Filter relevant files (files_containing(), files_importing())│
│  3. Recursively call sub-LLMs on targeted snippets (llm_query()) │
│  4. Build findings incrementally in variables                    │
│  5. Return final report via FINAL() or FINAL_VAR()              │
└─────────────────────────────────────────────────────────────────┘
```

## Research Paper Insights Applied

Based on analysis of the recursive language models paper (Khattab et al.), we identified several key learnings:

### 1. FINAL() brittleness is a known issue
The paper noted "distinguishing between a final answer and a thought is brittle" and they "added minor safeguards" but punted full resolution to future model training. Our implementation provides robust safeguards with escape hatches.

### 2. Specific character targets are effective
The paper's fix for Qwen making too many sub-LLM calls was adding: "aim for around ~200k characters per call" and noting "sub-LLMs can handle around 500K chars". We've incorporated these specific targets.

### 3. Models make non-optimal choices
The paper observed models "carefully construct their final answer... but then discard this information and continue wasting sub-calls". We've added detection for this wasteful pattern.

### 4. Efficiency framing works better than punishment
The paper framed batching guidance around "runtime costs" rather than rejection. We've updated our warning messages accordingly.

## Changes in This Iteration

- **Escalating prompts with efficiency-focused messaging** (not just rejection warnings)
- **Force-accept escape hatch** after max rejections to prevent infinite loops
- **Exploration score escape hatch** for legitimate edge cases
- **Scaled thresholds** based on repository size (logarithmic scaling for 100K to 10M+ token repos)
- **Agent behavior guardrails**:
  - Block filesystem imports (`pathlib`, `os`, `subprocess`, `shutil`)
  - Enforce sub-LLM usage for large repos
  - Iteration checkpoints with strategy nudges
  - Pre-populated first turn showing repo context
  - Tool usage metrics tracking
- **PR mode detection** via `get_changed_files()` from git diff
- **Review quality validation** ensuring actual code analysis, not architecture summaries
- **FINAL() auto-sanitization** converting problematic backtick patterns to FINAL_VAR()

## Changes Planned for Next Iteration

See `PLAN-rlm-fixes.md` for detailed task breakdown:

### Task 1.5: Wasteful Restart Pattern Detection
Detect when agent re-initializes variables or makes redundant queries that were already answered.

### Tasks 2.1-2.4: Improved Code Block Extraction
- Extract LAST code block (most likely to be executable)
- Syntax validation before execution
- String literal detection for edge cases

### Tasks 3.1-3.5: Enhanced Prompts
- Capacity guidelines (~500K chars/call)
- Tiered function ordering in prompt
- Query similarity tracking to detect redundant sub-LLM calls

## Architecture

```
cli.py (--rlm flag)
    │
    ▼
runner.py (_run_rlm_mode / run_rlm_review_with_fallback)
    │
    ▼
rlm/runner.py (RLMRunner orchestration loop)
    │
    ├──► get_review_prompt() ──► Comprehensive prompt with strategy guidance
    │
    ├──► _extract_code() ──► Smart extraction (skip examples, prefer FINAL)
    │    └──► _sanitize_final_call() ──► Auto-convert FINAL() with backticks
    │
    ├──► _validate_review_quality() ──► Check for severity markers, file refs
    │
    ├──► _build_continuation_prompt() ──► History + results + code-only enforcement
    │
    └──► REPL execution loop:
         │
         ▼
    rlm/repl.py (REPLProcess)
         │
         ├──► Import validation (blocks pathlib, os, subprocess, shutil)
         ├──► repo.files, repo.structure (codebase access)
         ├──► repo.changed_files (PR mode - auto-populated from git)
         ├──► llm_query(), llm_query_parallel() (sub-LLM calls)
         ├──► files_containing(), files_importing() (search)
         ├──► file_exists(), list_files_matching() (discovery)
         ├──► get_file_slice() (large file handling)
         └──► FINAL(), FINAL_VAR() (completion signals)
```

## Test Plan

- [ ] Unit tests for escalating prompts and threshold calculations
- [ ] Unit tests for code block extraction edge cases
- [ ] Integration test with mock LLM that doesn't comply
- [ ] Manual testing on small (<100K), medium (100K-1M), and large (1M+) repos

## Testing Status

```
============================= test session starts ==============================
124 passed in 34.37s
mypy: Success, no issues found in 17 source files
ruff: All checks passed!
```

## References

- [Recursive Language Models (Zhang et al., MIT CSAIL)](https://arxiv.org/abs/2503.09813)
- Test run analysis: `test_run.log` showing iteration patterns and efficiency improvements

## Checklist

- [x] Code follows project style guidelines
- [x] Self-review completed
- [x] Tests pass locally (124/124 passed)
- [x] Linting passes (ruff + mypy clean)
- [x] Documentation includes learnings from paper and implementation

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+# Install git-delta
+ARG GIT_DELTA_VERSION=0.18.2
+RUN ARCH=$(dpkg --print-architecture) && \
+  curl -fsSL "https://github.com/dandavison/delta/releases/download/${GIT_DELTA_VERSION}/git-delta_${GIT_DELTA_VERSION}_${ARCH}.deb" -o /tmp/git-delta.deb && \
+  dpkg -i /tmp/git-delta.deb && \
+  rm /tmp/git-delta.deb
+
+# Install uv (Python package manager) via multi-stage copy
+COPY --from=ghcr.io/astral-sh/uv:latest /uv /usr/local/bin/uv
```
[the remark this answers is not in the document]
Comment (anderskev) [reply .devcontainer/Dockerfile]: Not applicable - the `.devcontainer/Dockerfile` does not exist in this repository. The devcontainer configuration was removed in commit ca214b8 ("chore: remove local devcontainer config and test plan").

File: daydream/rlm/runner.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+        max_iterations: Maximum number of LLM iterations.
+    """
+
+    workspace_path: str
+    languages: list[str]
+    model: str = "opus"
+    sub_model: str = "haiku"
+    use_container: bool = True
+    pr_number: int | None = None
+    timeout: float = 600.0
```
[the remark this answers is not in the document]
Comment (anderskev) [reply daydream/rlm/runner.py]: @coderabbitai Acknowledged - `RLMConfig.timeout` is currently unused. This is intentional for the initial implementation which uses `max_iterations` as the primary bound. Wall-clock timeout enforcement can be added as an enhancement when needed.

File: daydream/rlm/environment.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    namespace["list_files_matching"] = list_files_matching
+
+    # File slice function
+    def get_file_slice(path: str, start_line: int, end_line: int) -> str:
+        """Get specific line range from a file (1-based, inclusive)."""
+        content = ctx.files.get(path, "")
+        lines = content.split("\n")
+        # Convert to 0-based index, end_line is inclusive
+        selected = lines[start_line - 1 : end_line]
+        return "\n".join(selected)
```
[the remark this answers is not in the document]
Comment (anderskev) [reply daydream/rlm/environment.py]: @coderabbitai Fixed. Added input validation and clamping in `get_file_slice()` to handle invalid line numbers gracefully - negative values are clamped to valid range and out-of-bounds requests return partial content instead of erroring.
